include/cuco/utility/fast_int.cuh

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * @brief Integer type with optimized division and modulo operators.
+ *
+ * @tparam T Underlying integer type
+ */
+template <typename T>
+struct fast_int {
+  static_assert(cuda::std::is_same_v<T, std::int32_t> or cuda::std::is_same_v<T, std::uint32_t>
+#if defined(CUCO_HAS_INT128)
+                  or cuda::std::is_same_v<T, std::int64_t> or cuda::std::is_same_v<T, std::uint64_t>
+#endif
+                ,
+                "Unsupported type");
+
+  using value_type = T;  ///< Underlying integer type
+
+  /**
+   * @brief Constructs a fast_int from an integer value.
+   *
+   * @param value Integer value
+   */
+  __host__ __device__ explicit constexpr fast_int(T value) noexcept : value_{value}
+  {
+    evaluate_magic_numbers();
+  }
+
+  /**
+   * @brief Explicit conversion operator to the underlying value type.
+   *
+   * @return Underlying value
+   */
+  __host__ __device__ explicit constexpr operator value_type() const noexcept { return value_; }
+
+ private:
+  using intermediate_type =
+    cuda::std::conditional_t<sizeof(value_type) == 4,
+                             std::uint64_t,
+                             unsigned __int128>;  ///< Intermediate type for multiplication
+  using unsigned_value_type = cuda::std::make_unsigned_t<value_type>;  ///< Unsigned value type
+  using signed_value_type   = cuda::std::make_signed_t<value_type>;    ///< Signed value type
+
+  static constexpr value_type value_bits =
+    CHAR_BIT * sizeof(value_type);  ///< Number of bits required to represent the value
+
+  /**
+   * @brief Computes the high bits of the multiplication of two unsigned integers.
+   *
+   * @param lhs Left-hand side of the multiplication
+   * @param rhs Right-hand side of the multiplication
+   *
+   * @return High bits of the multiplication
+   */
+  __host__ __device__ constexpr value_type mulhi(unsigned_value_type lhs,
+                                                 unsigned_value_type rhs) const noexcept
+  {
+#if defined(__CUDA_ARCH__)
+    if constexpr (sizeof(value_type) == 4) {
+      return __umulhi(lhs, rhs);
+    } else {
+      return __umul64hi(lhs, rhs);
+    }
+#else
+    return (intermediate_type(lhs) * intermediate_type(rhs)) >> value_bits;
+#endif
+  }
+
+  /**
+   * @brief Computes the log2 of an unsigned integer.
+   *
+   * @param v Unsigned integer
+   *
+   * @return Log2 of the unsigned integer
+   */
+  __host__ __device__ constexpr value_type log2(value_type v) const noexcept
+  {
+#if defined(__CUDA_ARCH__)
+    if constexpr (sizeof(value_type) == 4) {
+      return 31 - __clz(signed_value_type(v));
+    } else {
+      return 63 - __clzll(signed_value_type(v));
+    }
+#else
+    return cuda::std::bit_width(unsigned_value_type(v)) - 1;
+#endif
+  }
+
+  /**
+   * @brief Computes the magic numbers for the fast division.
+   */
+  __host__ __device__ constexpr void evaluate_magic_numbers() noexcept
+  {
+    // TODO assert(value_ > 0);
+    auto const val_log2 = log2(value_);
```

Comment:
```suggestion
    auto const val_log2 = this->log2(value_);
```
For readability, it's neither a CUDA instruction nor C/C++ math function but our local custom `log2`.

Reply:
Agreed. This way we make sure we call the member function.

Reply:
Forgot to mention, it would be great to add a benchmark in this PR for performance evaluation. Otherwise, showing a comparison between raw int and the fast int performance difference in the PR comment is sufficient.

Reply:
Agreed, I'd be curious to see how performance scales between `int` and `fast_int` with the size of the hash table.

Reply:
Yes, good idea. The perf benefit is most likely only visible if the map fits more or less into L2$. First, I must integrate this new util class into our existing extent design in #304. I will add benchmarks to that PR.

Reply:
> some artificial benchmarks that call operator% in a tight loop if this makes sense.

works for me

Reply:
The initial change suggestion of this thread has been incorporated in [9cf6bb3](https://github.com/NVIDIA/cuCollections/pull/309/commits/9cf6bb3abf65706fbfae262cdd2a383bb05b9f23). I'll mark this discussion as resolved.